// prerender-buildhooks/functions-bundled/nf-prerender-ext_utils/completion-detector.mts
// src/functions/utils/completion-detector.mts
var CompletionDetector = class {
  config;
  page;
  requestInterceptor;
  waitStart = 0;
  lastInFlightReportTime = 0;
  constructor(config, page, requestInterceptor) {
    this.config = config;
    this.page = page;
    this.requestInterceptor = requestInterceptor;
  }
  /**
   * Wait for page completion using configured strategies
   * Returns when either prerenderReady is true, network is idle, or timeout is reached
   */
  async waitForCompletion() {
    console.log("Waiting for window.prerenderReady or requests to settle...");
    this.waitStart = Date.now();
    this.lastInFlightReportTime = this.waitStart;
    while (true) {
      const result = await this.checkIfDone();
      if (result.completed) {
        return result;
      }
      await new Promise((resolve) => setTimeout(resolve, this.config.pageDoneCheckInterval));
    }
  }
  /**
   * Core completion detection logic
   * Evaluates prerenderReady flag and network activity
   */
  async checkIfDone() {
    const now = Date.now();
    const totalWaitTime = now - this.waitStart;
    if (totalWaitTime >= this.config.maxWaitTime) {
      return {
        completed: true,
        totalWaitTime,
        completionReason: "timeout"
      };
    }
    const currentPrerenderReady = await this.page.evaluate(() => {
      return typeof globalThis.prerenderReady === "boolean" ? globalThis.prerenderReady : null;
    });
    if (currentPrerenderReady !== null) {
      if (currentPrerenderReady === true) {
        return {
          completed: true,
          totalWaitTime,
          completionReason: "prerender-ready"
        };
      }
      return { completed: false, totalWaitTime, completionReason: "prerender-ready" };
    }
    const timeSinceLastRequest = now - this.requestInterceptor.getLastRequestTime();
    const numRequestsInFlight = this.requestInterceptor.getInFlightCount();
    const requestsSettled = numRequestsInFlight <= 0 && timeSinceLastRequest >= this.config.waitAfterLastRequest;
    if (numRequestsInFlight > 0 && totalWaitTime >= this.config.inFlightReportAfterTime) {
      if (now - this.lastInFlightReportTime > this.config.inFlightReportInterval) {
        const inflightUrls = this.requestInterceptor.getInFlightUrls();
        console.log(`In-flight requests after ${totalWaitTime}ms of wait: `, inflightUrls);
        this.lastInFlightReportTime = now;
      }
    }
    if (requestsSettled) {
      return {
        completed: true,
        totalWaitTime,
        completionReason: "network-idle"
      };
    }
    return { completed: false, totalWaitTime, completionReason: "network-idle" };
  }
};
export {
  CompletionDetector
};
