// prerender-buildhooks/functions-bundled/nf-prerender-ext_utils/performance-timer.mts
// src/functions/utils/performance-timer.mts
var PerformanceTimer = class {
  startTime;
  phases = /* @__PURE__ */ new Map();
  constructor() {
    this.startTime = Date.now();
  }
  /**
   * Start timing a named phase
   */
  startPhase(phaseName) {
    this.phases.set(phaseName, { start: Date.now() });
  }
  /**
   * End timing a named phase and calculate duration
   */
  endPhase(phaseName) {
    const phase = this.phases.get(phaseName);
    if (!phase) {
      throw new Error(`Phase '${phaseName}' was not started`);
    }
    const end = Date.now();
    const duration = end - phase.start;
    this.phases.set(phaseName, {
      ...phase,
      end,
      duration
    });
    return duration;
  }
  /**
   * Get duration of a completed phase
   */
  getPhaseDuration(phaseName) {
    const phase = this.phases.get(phaseName);
    if (!phase?.duration) {
      throw new Error(`Phase '${phaseName}' is not completed`);
    }
    return phase.duration;
  }
  /**
   * Get total elapsed time since timer creation
   */
  getTotalElapsed() {
    return Date.now() - this.startTime;
  }
  /**
   * Get all phase durations as an object
   */
  getAllPhaseDurations() {
    const result = {};
    for (const [name, phase] of this.phases) {
      if (phase.duration !== void 0) {
        result[name] = phase.duration;
      }
    }
    return result;
  }
};
export {
  PerformanceTimer
};
