// prerender-buildhooks/functions/nf-prerender-ext_utils/prerender-config.mts
// src/functions/utils/common-third-parties.mts
var COMMON_THIRD_PARTY_SKIP_LIST = [
  // Analytics & Tracking
  "2o7.net",
  "google-analytics.com",
  "googletagmanager.com",
  "amplitude.com",
  "mixpanel.com",
  "segment.com",
  "hotjar.com",
  "crazyegg.com",
  "luckyorange.com",
  "mouseflow.com",
  "fullstory.com",
  "clarity.ms",
  "newrelic.com",
  "nr-data.net",
  "datadoghq.com",
  "sentry.io",
  "quantummetric.com",
  "plausible.io",
  "heapanalytics.com",
  ".bugherd.com",
  ".bugsnag.com",
  ".pardot.com",
  "rudderstack.com",
  ".asortiz.com",
  ".mediavine.com",
  ".grow.me",
  // Advertising & Marketing, Payment, Monetization
  "adnxs.com",
  "adobedtm.com",
  "adroll.com",
  "ads-twitter.com",
  "adsafeprotected.com",
  "adsrvr.org",
  "adsymptotic.com",
  "advertising.com",
  "bing.com",
  "bizographics.com",
  "bkrtx.com",
  "bluekai.com",
  "casalemedia.com",
  "criteo.com",
  "criteo.net",
  "crwdcntrl.net",
  "dartsearch.net",
  "demdex.net",
  "doubleclick.net",
  "doubleverify.com",
  "everesttech.net",
  "facebook.com",
  "facebook.net",
  "fw-cdn.com",
  "googleadservices.com",
  "googlesyndication.com",
  "mathtag.com",
  "moatads.com",
  "monetate.net",
  "omtrdc.net",
  "openx.net",
  "optimizely.com",
  "outbrain.com",
  "pubmatic.com",
  "quantserve.com",
  "rubiconproject.com",
  "sc-static.net",
  "taboola.com",
  "tiqcdn.com",
  "visualwebsiteoptimizer.com",
  "yieldmanager.com",
  "mountain.com",
  "intentiq.com",
  "ad-delivery.net",
  ".ad.gt",
  "id5-sync.com",
  "nitropay.com",
  "braintreegateway.com",
  "pay.google.com",
  "play.google.com",
  "applepay.cdn-apple.com",
  "js.stripe.com",
  "www.ezojs.com",
  "ezoic.net",
  "chargebee.com",
  ".turn.com",
  "cdn.mxpnl.com",
  "onesignal.com",
  "paypal.com",
  "paypalobjects.com",
  ".adform.net",
  "razorpay.com",
  "formstack.com",
  // Social Media, Social Login
  "licdn.com",
  "snap.licdn.com",
  "pinimg.com",
  "pinterest.com",
  "redditstatic.com",
  "tiktok.com",
  "sharethis.com",
  "appleid.cdn-apple.com",
  "profitwell.com",
  "firstpromoter.com",
  "accounts.google.com",
  // Customer Support & Chat
  "crisp.chat",
  "driftt.com",
  "intercom.io",
  "intercomassets.com",
  "livechatinc.com",
  "tawk.to",
  "tidio.co",
  "zdassets.com",
  "zopim.com",
  "liveperson.net",
  // Forms & Surveys
  "hubspot.com",
  "hs-analytics.net",
  "hs-scripts.com",
  "hsforms.net",
  "qualified.com",
  "qualtrics.com",
  // Consent scripts, captchas, misc.
  ".iubenda.com",
  ".userway.org",
  "google.com/recaptcha",
  "cookielaw.org",
  "cookiepro.com",
  "cookiebot.com",
  "trustarc.com",
  "truste.com",
  "quantcast.com",
  "consensu.org",
  "usercentrics.eu",
  "osano.com",
  "privacy-center.org",
  "civiccomputing.com",
  "consentmanager.net",
  "privacymanager.io",
  "equalweb.com",
  "gatekeeperconsent.com",
  "cookie-script.com",
  "cdn-cookieyes.com",
  ".onetrust.com",
  // Media
  "www.youtube.com",
  "www.youtube-nocookie.com",
  ".live-video.net",
  // Netlify RUM
  "ingesteer.services-prod.nsvcs.net/rum_collection",
  // Real-time updates to data (these never stop...)
  "/google.firestore.v1beta1.Firestore/Listen/",
  "/google.firestore.v1.Firestore/Listen/",
  "events.launchdarkly.com"
];

// src/schema/prerender-configuration.ts
import { z } from "zod";
var urlSubstringValidator = z.string().optional().refine((value) => {
  if (!value) return true;
  const parts = value.split(",").map((part) => part.trim());
  return parts.every((part) => {
    const regexOnlyChars = /[[\](){}*+^$|\\]/;
    return !regexOnlyChars.test(part) && part.length > 0;
  });
}, {
  message: "Must be a comma-separated list of URL substrings (no regex characters like [], (), {}, *, +, ^, $, |, \\)"
});
var outputFormatValidator = z.string().default("html").refine((value) => {
  const validComponents = ["html", "tree", "report", "screenshot"];
  if (validComponents.includes(value)) {
    return true;
  }
  if (value.includes(",")) {
    const components = value.split(",").map((c) => c.trim());
    const uniqueComponents = new Set(components);
    return components.length === uniqueComponents.size && components.length > 0 && components.every((c) => validComponents.includes(c));
  }
  return false;
}, {
  message: "Must be 'html', 'tree', 'report', 'screenshot', or a comma-separated list like 'html,tree,report,screenshot'"
});
var prerenderConfigSchema = z.object({
  enabled: z.boolean().default(false),
  skipAgentsWithJS: z.boolean().default(true),
  userAgent: z.string().default("Netlify Prerender Function"),
  disableCaching: z.boolean().default(false),
  skipVisualResources: z.boolean().default(true),
  skipThirdParty: z.enum(["none", "all", "common"]).default("common"),
  skipAlways: urlSubstringValidator,
  skipNever: urlSubstringValidator,
  detailedReporting: z.boolean().default(false),
  cacheMaxAgeHours: z.number().min(6).max(720).default(72),
  // 6 hours to 30 days, default 3 days
  cacheSwrHours: z.number().min(0).max(168).default(24),
  // 0 to 7 days, default 1 day
  outputFormat: outputFormatValidator,
  forceNewBrowser: z.boolean().default(false),
  useBrowserCache: z.boolean().default(true)
});
var optionalPrerenderConfigSchema = prerenderConfigSchema.omit({ enabled: true, skipAgentsWithJS: true });
var optionalConfigDefaults = optionalPrerenderConfigSchema.parse({});
var BLOB_STORE_NAME = "netlify_prerender_extension";
var BLOB_OBJECT_NAME = "config.json";
function validateOptionalConfig(rawString) {
  try {
    if (!rawString || rawString.length === 0)
      return optionalConfigDefaults;
    const validatedConfig = optionalPrerenderConfigSchema.parse(JSON.parse(rawString));
    return validatedConfig;
  } catch (e) {
    console.error(`Error parsing optional config: [${rawString}], error:`, e);
    return optionalConfigDefaults;
  }
}
function optionalConfigWithoutDefaults(input) {
  function getObjectDifference(o1, o2) {
    const diff = {};
    for (const key in o2) {
      if (o1[key] !== o2[key]) {
        diff[key] = o2[key];
      }
    }
    return diff;
  }
  return getObjectDifference(optionalConfigDefaults, input);
}

// src/functions/utils/prerender-config.mts
import { getStore } from "@netlify/blobs";
var THIRD_PARTY_SKIP_MODES = ["none", "all", "common"];
var OUTPUT_COMPONENTS = ["html", "tree", "report", "screenshot"];
function parseOutputFormat(formatString) {
  if (formatString.includes(",")) {
    return formatString.split(",").map((c) => c.trim());
  }
  return [formatString];
}
async function createPrerenderConfig(req) {
  const isProduction = !process.env.NETLIFY_DEV;
  const allowRemoteHosts = "true" === process.env.NETLIFY_PRERENDER_ALLOW_REMOTE_HOSTS?.trim().toLowerCase();
  if (allowRemoteHosts) {
    console.warn(`==> NOTE: allowRemoteHosts IS ON`);
  }
  const rawDirectConfig = req.headers.get("X-Prerender-Direct-Config");
  const optionalConfig = await loadOptionalConfig(rawDirectConfig);
  const config = {
    directlyInvoked: rawDirectConfig !== null,
    /* Always read from dedicated env vars */
    isProduction,
    allowRemoteHosts,
    /* Optional config */
    userAgent: optionalConfig.userAgent,
    disableCaching: optionalConfig.disableCaching,
    cacheMaxAgeHours: optionalConfig.cacheMaxAgeHours,
    cacheSwrHours: optionalConfig.cacheSwrHours,
    enableDetailedReporting: optionalConfig.detailedReporting,
    skipVisualResources: optionalConfig.skipVisualResources,
    skipThirdParty: THIRD_PARTY_SKIP_MODES.includes(optionalConfig.skipThirdParty) ? optionalConfig.skipThirdParty : "common",
    skipAlways: optionalConfig.skipAlways?.split(",").filter(Boolean) || [],
    skipNever: optionalConfig.skipNever?.split(",").filter(Boolean) || [],
    outputFormats: parseOutputFormat(optionalConfig.outputFormat),
    forceNewBrowser: optionalConfig.forceNewBrowser,
    // Don't allow browser cache when allowing multiple domains, to avoid any cache attacks
    useBrowserCache: !allowRemoteHosts && optionalConfig.useBrowserCache,
    /* Currently hard-coded settings below */
    commonThirdPartySkipList: COMMON_THIRD_PARTY_SKIP_LIST,
    visualResourceTypes: ["image", "media", "font", "stylesheet"],
    skipSSE: true,
    waitAfterLastRequest: 500,
    pageDoneCheckInterval: 100,
    maxWaitTime: 15e3,
    inFlightReportAfterTime: 1e3,
    inFlightReportInterval: 1e3,
    navigationTimeout: 15e3,
    useAsyncCleanup: false
    // Currently disable this optimization, to avoid hard-to-debug async cleanup failures
  };
  return config;
}
async function loadOptionalConfig(rawDirectConfig) {
  let result = optionalConfigDefaults;
  const rawConfigFromEnv = process.env.NETLIFY_PRERENDER_CONFIG;
  if (rawConfigFromEnv) {
    console.log("Using optional configuration from env var");
    result = validateOptionalConfig(rawConfigFromEnv);
  } else {
    try {
      const store = getStore(BLOB_STORE_NAME);
      const rawConfigFromBlob = await store.get(BLOB_OBJECT_NAME, { consistency: "strong" });
      if (rawConfigFromBlob) {
        console.log("Using optional configuration from blob storage");
        result = validateOptionalConfig(rawConfigFromBlob);
      }
    } catch (error) {
      console.error("Failed to load from blob storage", error);
    }
  }
  if (rawDirectConfig !== null) {
    const addedConfig = optionalConfigWithoutDefaults(validateOptionalConfig(rawDirectConfig));
    console.log("Function directly invoked, caching disabled");
    result = { ...result, ...addedConfig, disableCaching: true };
  }
  console.log("Optional config: ", JSON.stringify(result, null, 2).replace(/\n\s*/g, " "));
  return result;
}
export {
  OUTPUT_COMPONENTS,
  THIRD_PARTY_SKIP_MODES,
  createPrerenderConfig,
  parseOutputFormat
};
