// prerender-buildhooks/functions/nf-prerender-ext_utils/request-validation.mts
// src/functions/utils/request-validation.mts
import { parse } from "psl";
import { isIPv4, isIPv6 } from "node:net";
var MAX_URL_LENGTH = 2048;
var MIN_TOKEN_LENGTH = 16;
function errorResult(message, status) {
  return {
    isValid: false,
    error: { message, status }
  };
}
function validateAuthToken(req) {
  const authToken = process.env.NETLIFY_PRERENDER_AUTH_TOKEN?.trim();
  if (!authToken || authToken.length < MIN_TOKEN_LENGTH) {
    const error = new Error(
      `NETLIFY_PRERENDER_AUTH_TOKEN is required and must be at least 16 bytes long. Current length: ${authToken?.length || 0} bytes`
    );
    console.error(error.message);
    return new Response("Configuration error", { status: 500 });
  }
  const providedToken = req.headers.get("X-Prerender-Token");
  if (!providedToken || providedToken.length < MIN_TOKEN_LENGTH || providedToken !== authToken) {
    const message = "Authentication token is " + (!providedToken ? "missing" : "invalid");
    return new Response(message, { status: 401 });
  }
  return null;
}
function validateTargetUrl(req, config) {
  const requestUrl = new URL(req.url);
  const targetUrlParam = requestUrl.searchParams.get("url");
  if (!targetUrlParam)
    return errorResult("Missing url parameter", 400);
  if (targetUrlParam.length > MAX_URL_LENGTH)
    return errorResult("url too long", 400);
  let parsedTargetUrl;
  try {
    parsedTargetUrl = new URL(targetUrlParam);
  } catch {
    return errorResult(`Invalid URL format: ${targetUrlParam}`, 400);
  }
  const domain = extractDomain(parsedTargetUrl.hostname);
  if (!domain)
    return errorResult(`Invalid URL domain: ${targetUrlParam}`, 400);
  if (!config.allowRemoteHosts) {
    if (parsedTargetUrl.host !== requestUrl.host || parsedTargetUrl.protocol !== requestUrl.protocol)
      return errorResult("Invalid target URL: must be same host & protocol", 403);
  }
  return {
    isValid: true,
    urlString: targetUrlParam,
    url: parsedTargetUrl,
    domain
  };
}
function extractDomain(hostname) {
  if (isIPv4(hostname) || isIPv6(hostname)) {
    return null;
  }
  if (hostname === "localhost") {
    return "localhost";
  }
  const result = parse(hostname);
  if (!("domain" in result) || !result.domain) {
    return null;
  }
  return result.domain;
}
export {
  extractDomain,
  validateAuthToken,
  validateTargetUrl
};
