// prerender-buildhooks/functions-bundled/nf-prerender-ext_prerender-keep-warm.mts
// src/functions/prerender-keep-warm.mts
var prerender_keep_warm_default = async () => {
  console.log("Keep-warm function triggered");
  try {
    const url = Netlify.env.get("URL");
    const authToken = Netlify.env.get("NETLIFY_PRERENDER_AUTH_TOKEN");
    const keepWarmCount = Netlify.env.get("NETLIFY_PRERENDER_KEEP_WARM");
    if (!url) {
      console.error("URL environment variable is not set");
      return;
    }
    if (!authToken) {
      console.error("NETLIFY_PRERENDER_AUTH_TOKEN environment variable is not set");
      return;
    }
    const count = parseInt(keepWarmCount || "1", 10);
    if (isNaN(count) || count < 1 || count > 10) {
      console.error(`Invalid NETLIFY_PRERENDER_KEEP_WARM value: ${keepWarmCount}. Must be between 1 and 10.`);
      return;
    }
    const prerenderUrl = `${url}/netlify-prerender-function`;
    const fetchPromises = Array.from(
      { length: count },
      () => fetch(prerenderUrl, {
        method: "GET",
        headers: {
          "X-Prerender-Token": authToken,
          "X-Prerender-Ping": "1"
        }
      })
    );
    const results = await Promise.allSettled(fetchPromises);
    const successCount = results.filter((result) => {
      if (result.status === "fulfilled" && result.value.ok) {
        return true;
      }
      return false;
    }).length;
    console.log(`Keep-warm ping completed: ${successCount}/${count} successful`);
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    console.error(`Keep-warm function error: ${errorMessage}`);
  }
};
var config = {
  schedule: "*/3 * * * *",
  // Run every 3 minutes
  generator: "netlify-prerender-extension",
  name: "Netlify Prerender Keep Warm"
  // name attribute is valid, but not yet exposed in type.
};
export {
  config,
  prerender_keep_warm_default as default
};
